irohad/validation/impl/stateful_validator_impl.cpp

```diff
@@ -53,41 +73,54 @@ namespace iroha {
       // Filter only valid transactions
       auto filter = [&temporaryWsv, checking_transaction](auto &acc,
                                                           const auto &tx) {
-        auto answer = temporaryWsv.apply(tx, checking_transaction);
+        auto answer =
+            temporaryWsv.apply(*tx->makeOldModel(), checking_transaction);
         if (answer) {
           acc.push_back(tx);
         }
         return acc;
       };
 
-      auto &txs = proposal.transactions;
+      auto &txs = proposal->transactions();
       decltype(txs) valid = {};
 
-      model::Proposal validated_proposal(
-          std::accumulate(txs.begin(), txs.end(), valid, filter));
-      validated_proposal.height = proposal.height;
+      auto valid_txs = std::accumulate(txs.begin(), txs.end(), valid, filter);
+
+      // rework validation logic, so that this conversion is not needed
+      auto valid_proto_txs =
+          valid_txs
+          | boost::adaptors::transformed([](const auto &polymorphic_tx) {
+              return *static_cast<const shared_model::proto::Transaction *>(
+                  polymorphic_tx.operator->());
+            });
+      auto validated_proposal = proto::ProposalBuilder()
+                                    .createdTime(proposal->created_time())
+                                    .height(proposal->height())
+                                    .transactions(valid_proto_txs)
+                                    .build();
+
       log_->info("transactions in verified proposal: {}",
-                 validated_proposal.transactions.size());
-      return validated_proposal;
+                 validated_proposal.transactions().size());
+      return shared_model::detail::makePolymorphic<decltype(
+          validated_proposal)>(validated_proposal.getTransport());
     }
 
     bool StatefulValidatorImpl::signaturesSubset(
-        const model::Transaction::SignaturesType &signatures,
-        const std::vector<pubkey_t> &public_keys) {
+        const shared_model::interface::Transaction::SignatureSetType
+            &signatures,
+        const std::vector<crypto::PublicKey> &public_keys) {
       // TODO 09/10/17 Lebedev: simplify the subset verification IR-510
       // #goodfirstissue
-      std::set<pubkey_t> txPubkeys;
+      std::unordered_set<std::string> txPubkeys;
       for (auto sign : signatures) {
-        txPubkeys.insert(sign.pubkey);
-      }
-      std::set<pubkey_t> accPubkeys;
-      for (auto pubkey : public_keys) {
-        accPubkeys.insert(pubkey);
+        txPubkeys.insert(sign->publicKey().toString());
       }
-      return std::includes(accPubkeys.begin(),
-                           accPubkeys.end(),
-                           txPubkeys.begin(),
-                           txPubkeys.end());
+      return std::all_of(public_keys.begin(),
+                         public_keys.end(),
+                         [&txPubkeys](const auto &public_key) {
```

Comment:
There's comparator for `std::includes` as well, use it. It more intuitive (and faster) to use it here

Reply:
Not suitable here. std::includes work with ordered sets. In order to get ordered sets from passed containers we will need to iterate through both containers and then perform `std::includes`. Using my approach we iterate through 1st container, and while we are iterating through second we return bool result

Reply:
It's only the matter of performance?

Reply:
It's only the matter of performance?